src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionTemplateComparator.kt

```diff
@@ -0,0 +1,206 @@
+package uk.gov.justice.digital.hmpps.visitscheduler.utils
+
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.visitscheduler.dto.sessions.SessionDetailsDto
+import uk.gov.justice.digital.hmpps.visitscheduler.dto.sessions.location.PermittedSessionLocationDto
+import uk.gov.justice.digital.hmpps.visitscheduler.model.entity.session.category.PrisonerCategoryType
+import uk.gov.justice.digital.hmpps.visitscheduler.model.entity.session.incentive.IncentiveLevel
+import java.util.*
+import java.util.function.BiPredicate
+import java.util.stream.Collectors
+
+@Component
+class SessionTemplateComparator(private val sessionDatesUtil: SessionDatesUtil) {
+
+  private val validMatch = object : BiPredicate<String?, String?> {
+    override fun test(permittedSessionLevel: String?, prisonerLevel: String?): Boolean {
+      permittedSessionLevel?.let {
+        return it == prisonerLevel
+      }
+      // If no prison level then match
+      return true
+    }
+  }
+
+  private val multipleLocationMatcher =
+    BiPredicate<PermittedSessionLocationDto, Set<PermittedSessionLocationDto>> { primarySessionLocation, comparedSessionLocations ->
+      comparedSessionLocations.stream().anyMatch { comparedSessionLocation ->
+        validMatch.test(primarySessionLocation.levelOneCode, comparedSessionLocation.levelOneCode)
+          .and(validMatch.test(primarySessionLocation.levelTwoCode, comparedSessionLocation.levelTwoCode))
+          .and(validMatch.test(primarySessionLocation.levelThreeCode, comparedSessionLocation.levelThreeCode))
+          .and(validMatch.test(primarySessionLocation.levelFourCode, comparedSessionLocation.levelFourCode))
+      }
+    }
+
+  private fun hasOverlappingDates(
+    newSessionDetails: SessionDetailsDto,
+    existingSessionDetails: SessionDetailsDto,
+  ): Boolean {
+    val newFromDate = newSessionDetails.sessionDateRange.validFromDate
+    val existingFromDate = existingSessionDetails.sessionDateRange.validFromDate
+    val newToDate = newSessionDetails.sessionDateRange.validToDate ?: newFromDate.plusYears(500)
+    val existingToDate = existingSessionDetails.sessionDateRange.validToDate ?: newSessionDetails.sessionDateRange.validToDate?.plusDays(1) ?: newToDate
+
```

Comment:
Not sure about this line trying to come up with a better way of doing this
` val existingToDate = existingSessionDetails.sessionDateRange.validToDate ?: newSessionDetails.sessionDateRange.validToDate?.plusDays(1) ?: newToDate
`
If there is no existing To date should it just assume it's an overlap?

Reply:
as discussed was tricky with the null so setting newDate to plus 500 years if its null and existingToDate to the same to avoid adding further if statements.